In [10]:
import pandas as pd
from sklearn.metrics import confusion_matrix
import sklearn.metrics

1.  Create a new file named model_evaluation.py or model_evaluation.ipynb for these exercises.

2.  Given the following confusion matrix, evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

### Accuracy = $\frac{TP + TN}{TP +TN +FP + FN}$ = $\frac{46 + 34}{46 + 34 + 13 + 7}$ = $\frac{4}{5}$ 
### Precision = $\frac{TP}{TP + FP}$ = $\frac{46}{46 + 13}$ = $\frac{46}{59}$
### Recall = $\frac{TP}{TP + FN}$ = $\frac{46}{46 + 7}$ = $\frac{46}{55}$
### F1 = $\frac{1}{2*(\frac{1}{Precision}+\frac{1}{Recall})}$ = $\frac{2 * Precision * Recall}{Precision + Recall}$ = $\frac{2* \frac{46}{59} * \frac{46}{55}}{\frac{46}{59} + \frac{46}{55}}$  

In the context of this problem, what is a false positive?

### predicting dog for an actual cat

In the context of this problem, what is a false negative?

### predicting cat for an actual dog

How would you describe this model?

Its accurate
Its precise
Its got good recal

3.  You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.
Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.
Use the predictions dataset and pandas to help answer the following questions:

In [2]:
c3 = pd.read_csv('c3.csv')
c3

actual     model1     model2     model3
0    No Defect  No Defect     Defect  No Defect
1    No Defect  No Defect     Defect     Defect
2    No Defect  No Defect     Defect  No Defect
3    No Defect     Defect     Defect     Defect
4    No Defect  No Defect     Defect  No Defect
..         ...        ...        ...        ...
195  No Defect  No Defect     Defect     Defect
196     Defect     Defect  No Defect  No Defect
197  No Defect  No Defect  No Defect  No Defect
198  No Defect  No Defect     Defect     Defect
199  No Defect  No Defect  No Defect     Defect

[200 rows x 4 columns]

An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

### True Positive:  BROKEN DUCK
### True Negative:  GOOD DUCK
### False Positive:  Predict a broken duck and actually getting a good duck
### False Negative:  Predict good duck and actually getting a bad duck

### We want to catch as many of the broken ducks as possible so we would choose to maximize recall

In [3]:
print(confusion_matrix(c3.actual, c3.model1, labels = ('Defect', 'No Defect')))

[[  8   8]
 [  2 182]]


In [4]:
print(confusion_matrix(c3.actual, c3.model2, labels = ('Defect', 'No Defect')))

[[  9   7]
 [ 81 103]]


In [5]:
print(confusion_matrix(c3.actual, c3.model3, labels = ('Defect', 'No Defect')))

[[13  3]
 [86 98]]


### We want to use the third model in order to maximize recall of $\frac{13}{16}$ where the other two give $\frac{1}{2}$ and $\frac{9}{16}$ respectively

Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

Under the key wording of the question you are trying to minimize false positives i.e going by the keywords they need you to predict which have defects but...doesn't have a defect.  The phrasing of this question is in direct to the part you dont want to accidentally give out a vacation package?  If you want to minimize vacation packages given out accidental and otherwise this is a conditional probability exercise.  You would rely on recall and throw out all predicted negatives so by anything approaching logic or real world you would still use recall and the last model. Throwing out 99 ducks predicted as defective...86 of them still good is still cheaper than one extra vacation package.  

4.  You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).
At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).
Several models have already been developed with the data, and you can find their results here.
Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [55]:
gives_you_paws = pd.read_csv('gives_you_paws.csv')
gives_you_paws

actual model1 model2 model3 model4
0       cat    cat    dog    cat    dog
1       dog    dog    cat    cat    dog
2       dog    cat    cat    cat    dog
3       dog    dog    dog    cat    dog
4       cat    cat    cat    dog    dog
...     ...    ...    ...    ...    ...
4995    dog    dog    dog    dog    dog
4996    dog    dog    cat    cat    dog
4997    dog    cat    cat    dog    dog
4998    cat    cat    cat    cat    dog
4999    dog    dog    dog    dog    dog

[5000 rows x 5 columns]

a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?  

In [28]:
gives_you_paws.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [24]:
3254/(3254+1746)

0.6508

In [25]:
gives_you_paws = gives_you_paws['baseline_prediction'] = 'dog'

TypeError: 'str' object does not support item assignment

In [29]:
print(confusion_matrix(gives_you_paws.actual, gives_you_paws.model1, labels = ('dog', 'cat')))

[[2614  640]
 [ 323 1423]]


This model gives an accuracy of $\frac{4037}{5000}$=.8074 which is very good and the best based on accuracy alone.

In [30]:
print(confusion_matrix(gives_you_paws.actual, gives_you_paws.model2, labels = ('dog', 'cat')))

[[1597 1657]
 [ 191 1555]]


This model gives an accuracy of $\frac{3152}{5000}$=.6304 which is below the baseline.

In [31]:
print(confusion_matrix(gives_you_paws.actual, gives_you_paws.model3, labels = ('dog', 'cat')))

[[1655 1599]
 [ 853  893]]


This model gives an accuracy of $\frac{2548}{5000}$=.5096 which is significantly below the baseline.


In [32]:
print(confusion_matrix(gives_you_paws.actual, gives_you_paws.model4, labels = ('dog', 'cat')))

[[3110  144]
 [1143  603]]


This last model gives an accuracy above the baseline with $\frac{3713}{5000}$ =.7426 

b.  Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?

Model 4 since it gives the best recall of $\frac{3110}{3254}$=.9557467732022127

c.  Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?

I would want model 2 as it gives the best specificity of $\frac{1555}{1746}$=.8906071019473081

5.  Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.
sklearn.metrics.accuracy_score
sklearn.metrics.precision_score
sklearn.metrics.recall_score
sklearn.metrics.classification_report

In [42]:
print(f'accuracy = {sklearn.metrics.accuracy_score(gives_you_paws.actual, gives_you_paws.model1)}')
print(f'precision = {sklearn.metrics.precision_score(gives_you_paws.actual, gives_you_paws.model1, pos_label="dog")}')
print(f'recall = {sklearn.metrics.recall_score(gives_you_paws.actual, gives_you_paws.model1, pos_label="dog")}')

accuracy = 0.8074
precision = 0.8900238338440586
recall = 0.803318992009834


In [49]:
print(f'{sklearn.metrics.classification_report(gives_you_paws.actual, gives_you_paws.model1)}')

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [44]:
print(f'accuracy = {sklearn.metrics.accuracy_score(gives_you_paws.actual, gives_you_paws.model2)}')
print(f'precision = {sklearn.metrics.precision_score(gives_you_paws.actual, gives_you_paws.model2, pos_label="dog")}')
print(f'recall = {sklearn.metrics.recall_score(gives_you_paws.actual, gives_you_paws.model2, pos_label="dog")}')

accuracy = 0.6304
precision = 0.8931767337807607
recall = 0.49078057775046097


In [50]:
print(f'{sklearn.metrics.classification_report(gives_you_paws.actual, gives_you_paws.model2)}')

              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000



In [45]:
print(f'accuracy = {sklearn.metrics.accuracy_score(gives_you_paws.actual, gives_you_paws.model3)}')
print(f'precision = {sklearn.metrics.precision_score(gives_you_paws.actual, gives_you_paws.model3, pos_label="dog")}')
print(f'recall = {sklearn.metrics.recall_score(gives_you_paws.actual, gives_you_paws.model3, pos_label="dog")}')

accuracy = 0.5096
precision = 0.6598883572567783
recall = 0.5086047940995697


In [52]:
print(f'{sklearn.metrics.classification_report(gives_you_paws.actual, gives_you_paws.model3)}')

              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000



In [46]:
print(f'accuracy = {sklearn.metrics.accuracy_score(gives_you_paws.actual, gives_you_paws.model)}')
print(f'precision = {sklearn.metrics.precision_score(gives_you_paws.actual, gives_you_paws.model4, pos_label="dog")}')
print(f'recall = {sklearn.metrics.recall_score(gives_you_paws.actual, gives_you_paws.model4, pos_label="dog")}')

accuracy = 0.7426
precision = 0.7312485304490948
recall = 0.9557467732022127


In [51]:
print(f'{sklearn.metrics.classification_report(gives_you_paws.actual, gives_you_paws.model4)}')

              precision    recall  f1-score   support

         cat       0.81      0.35      0.48      1746
         dog       0.73      0.96      0.83      3254

    accuracy                           0.74      5000
   macro avg       0.77      0.65      0.66      5000
weighted avg       0.76      0.74      0.71      5000

